### PR TITLE
feat(actuators): add GRU and GRU-residual network actuator models

### PR DESCRIPTION
**Pull request by vidurv-nvidia**
# Description

Adds two recurrent (GRU) actuator network models to `isaaclab.actuators`, alongside the existing MLP and LSTM models:

- **`ActuatorNetGRU`** — explicit actuator; a GRU predicts the *total* joint effort from the joint position error and velocity. Derives from `IdealPDActuator`, so the predicted effort is clipped to the actuator's effort limit.
- **`ActuatorNetGRUResidual`** — implicit-PD actuator that adds a GRU-predicted *residual* feed-forward effort. The physics engine applies the PD term using the configured stiffness/damping; the desired positions/velocities are preserved on return.

Both load a TorchScript network that exposes a `.gru` submodule and implements `forward(x, hidden) -> (output, hidden)` with a 2-dim input `[position_error, velocity]` (mirroring `ActuatorNetLSTM`). Each input and the output support optional `(mean, std)` normalization via config fields (`None` = identity). A non-finite network output is sanitized to zero effort before it reaches the engine.

These run on the standard (torch) actuator path. The Newton native-actuator path (`use_newton_actuators=True`) does not yet provide a GRU controller, so these configs are not authored onto that path; adding a Newton-side GRU controller would be a separate change.

Fixes # (no associated issue)

## Type of change

- New feature (non-breaking change which adds functionality)

## Screenshots

N/A — no visual change.

## Checklist

- [x] I have read and understood the [contribution guidelines](https://isaac-sim.github.io/IsaacLab/main/source/refs/contributing.html)
- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added a changelog fragment under `source/<pkg>/changelog.d/` for every touched package (do **not** edit `CHANGELOG.rst` or bump `extension.toml` — CI handles that)
- [x] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

### CODE REVIEW

File: source/isaaclab/isaaclab/actuators/actuator_net.py [hunk 1]
```diff
@@ -98,6 +104,209 @@ def compute(
...
+
+    def __init__(self, cfg: ActuatorNetGRUCfg, *args, **kwargs):
+        super().__init__(cfg, *args, **kwargs)
+        self._init_gru_runtime()
+
+    """
+    Operations.
+    """
+
+    def reset(self, env_ids: Sequence[int]):
```
[the remark this answers is not in the document]
Comment (vidurv-nvidia) [reply source/isaaclab/isaaclab/actuators/actuator_net.py:273]: Done in 63903b9 — `ActuatorNetGRU.reset()` now calls `super().reset(env_ids)` before resetting the hidden state, matching `ActuatorNetGRUResidual`.

File: source/isaaclab/isaaclab/actuators/actuator_net.py [hunk 2]
```diff
@@ -98,6 +104,209 @@ def compute(
...
+        self.sea_input = torch.zeros(batch, 1, 2, device=self._device)
+        self.sea_hidden_state = torch.zeros(num_layers, batch, hidden_dim, device=self._device)
+        # per-env view for resets (shares storage)
+        self.sea_hidden_state_per_env = self.sea_hidden_state.view(
+            num_layers, self._num_envs, self.num_joints, hidden_dim
+        )
+
+    def _resolve_normalization(self, stats: tuple[float, float] | None) -> tuple[float, float]:
+        """Return the ``(mean, std)`` to apply, defaulting to identity and flooring the std."""
+        if stats is None:
```
[the remark this answers is not in the document]
Comment (vidurv-nvidia) [reply source/isaaclab/isaaclab/actuators/actuator_net.py:179]: Done in 63903b9 — `_resolve_normalization` now logs a warning when a configured `std` is below the floor (and identifies which normalization field), since a near-zero std amplifies the normalized values.

File: source/isaaclab/isaaclab/actuators/actuator_net.py [hunk 3]
```diff
@@ -98,6 +107,227 @@ def compute(
...
+        if control_action.joint_positions is None:
+            raise ValueError("GRU actuator input requires control_action.joint_positions to be set.")
+        # normalized [position_error, velocity] inputs
+        pos_error = (control_action.joint_positions - joint_pos).flatten()
+        self.sea_input[:, 0, 0] = (pos_error - self._pos_error_norm[0]) / self._pos_error_norm[1]
+        self.sea_input[:, 0, 1] = (joint_vel.flatten() - self._vel_norm[0]) / self._vel_norm[1]
+
+        # run inference, then denormalize and guard against a non-finite output
+        with torch.inference_mode():
+            output, self.sea_hidden_state[:] = self.network(self.sea_input, self.sea_hidden_state)
```
[the remark this answers is not in the document]
Comment (vidurv-nvidia) [reply source/isaaclab/isaaclab/actuators/actuator_net.py:236]: Fixed in e1dea0f — `_resolve_normalization` now raises `ValueError` for a negative `std` (only `0 ≤ std < 1e-8` is floored, with the warning). Added a regression test (`test_actuator_net_gru_negative_std_raises`).

File: source/isaaclab/isaaclab/actuators/actuator_net.py [hunk 4]
```diff
@@ -98,6 +107,227 @@ def compute(
...
+        self._init_gru_runtime()
+
+    """
+    Operations.
+    """
+
+    def reset(self, env_ids: Sequence[int]):
+        super().reset(env_ids)
+        self._reset_gru_state(env_ids)
+
```
[the remark this answers is not in the document]
Comment (vidurv-nvidia) [reply source/isaaclab/isaaclab/actuators/actuator_net.py:276]: Fixed in e1dea0f — the denormalization and `nan_to_num` now run inside the `with torch.inference_mode()` block.

**Comment by AntoineRichard** [comment]
Hi @vidurv-nvidia could you target newton's actuators? We will be deprecating our internal actuators, I think it would make sense to use the shared actuators they built this way IsaacSim and Newton could use them too. Or at least could these actuators modify the newton actuators rather than our own? @jvonmuralt for viz.

**Comment by vidurv-nvidia** [comment]
Closing as stale. This branch is ~970 commits behind `develop` and has not been updated since June; the actuator APIs it builds on have changed substantially since. Happy to revive it as a fresh branch against current `develop` if there is still demand for GRU actuator models.
